### PR TITLE
[CenterOfMass] Searchbar, BOM import, Zero mass, Fixes

### PR DESCRIPTION
**Pull request by s-quirin**
- [x] Please check this box if you're not submitting a new macro.   

---

### Changelog:
- **New**: Searchbar for solids
- **New**: Reworked import function to improve external bill of materials (BOM) import with better input tolerance
- **Fix**: Changed caption "weight" to "mass" in export file
- **Code**: Indentation of on_pushButton_Export() similar to on_pushButton_Import()
- **Code**: Idiomatic python
- **New**: Mass can be set zero to exclude solid from calculation and visualization
- **Fix**: Behaviour of default value spin and "Apply to all"
- **Fix**: Preserve original solids color when "New" button pressed

### CODE REVIEW

File: Information/CenterOfMass.FCMacro [hunk 1]
```diff
@@ -795,23 +824,33 @@ class CenterofmassWidget(QtGui.QWidget):
             progress_bar.next()
         progress_bar.stop()
 
-    def compute_centerOfMass(self):
+    def compute_centerOfMass(self, setAllow=None):
         """Compute joint center of mass from all objects."""
```
Comment (galou) [inline Information/CenterOfMass.FCMacro:828]: I have difficulties imagining what `setAllow`can mean. Would it be possible to rename the parameter or add an explanation in the DocString?

File: Information/CenterOfMass.FCMacro [hunk 2]
```diff
@@ -1017,80 +1059,121 @@ class CenterofmassWidget(QtGui.QWidget):
...
-                f.close()
-                app.Console.PrintMessage(fileName + ' saved \n')
-            except:
-                error_message('Error writing file ' + fileName)
+            app.Console.PrintWarning('No file saved \n')
+            return
+        app.Console.PrintMessage('Saving ' + fileName + '\n')
+        densTot = self.massTot/self.volTot if self.volTot != 0 else 0.
+        head = ['Number', 'Label', 'Material',
+                'Volume (%s)'%self.unitForV,
```
Comment (galou) [inline Information/CenterOfMass.FCMacro:1073]: Please prefer f-strings `f'Volume ({self.unitForV})'`.

File: Information/CenterOfMass.FCMacro [hunk 3]
```diff
@@ -1017,80 +1059,121 @@ class CenterofmassWidget(QtGui.QWidget):
...
+                'Density (%s)'%self.unitForD,
+                'Mass (%s)'%self.unitForM,
+              *('Center of mass ' + axis + ' (' + self.unitForL + ')' for axis in ['X','Y','Z'])]
+        foot = ['Total', '', '',
+                '%.6e'%self.volTot,
```
Comment (galou) [inline Information/CenterOfMass.FCMacro:1078]: Here the f-string would be `f'{self.volTot:.6e}'`.

File: Information/CenterOfMass.FCMacro [hunk 4]
```diff
@@ -1017,80 +1059,121 @@ class CenterofmassWidget(QtGui.QWidget):
...
+            return
+        if selectedFilter == fileFilter[-1]:
+            import locale
+            encoding = locale.getpreferredencoding()
+            delimiter = ';'
+        with open(fileName, 'r', encoding=encoding) as csvfile:
+            app.Console.PrintMessage('Reading from %s\n' % fileName)
+            reader = csv.DictReader(csvfile, delimiter=delimiter, skipinitialspace=True)
+            col = {}    # first columns for label(s), density/ies, mass(es), material(s)
+            for name in ['label', 'densit', 'mass', 'material']:
```
Comment (galou) [inline Information/CenterOfMass.FCMacro:1116]: What is the advantage of using `densit` over the more clear `density`?

File: Information/CenterOfMass.FCMacro [hunk 5]
```diff
@@ -1017,80 +1059,121 @@ class CenterofmassWidget(QtGui.QWidget):
...
+                col[name] = next((s for s in reader.fieldnames if name in s.lower()), None)
+            if col['mass'] and 'cent' in col['mass'].lower():
+                col['mass'] = None    # col['mass'] was "Center of mass"
```
Comment (galou) [inline Information/CenterOfMass.FCMacro:1119]: Please remove the comment.

**Review by galou — APPROVED** [review]
Thanks for your contribution!

**Comment by galou** [comment]
@chupins Are the changes OK for you?

File: Information/CenterOfMass.FCMacro (hunk 4 above)
Comment (s-quirin) [reply Information/CenterOfMass.FCMacro:1116]: These are the search strings covering singular, plural and compound words

File: Information/CenterOfMass.FCMacro (hunk 5 above)
Comment (s-quirin) [reply Information/CenterOfMass.FCMacro]: Comment has been clarified

**Comment by chupins** [comment]
Completly ok for me.
Thanks @s-quirin for your answer and polishing !

**Comment by luzpaz** [comment]
bump

**Pull request merged**